### PR TITLE
Suggest borrowing when it would satisfy an unmet trait bound

### LINKED ISSUE #56368: E0277 error is confusing when compile fails using the last implementor for a trait
## Description

This is a followup of forum post [[Solved] Need help understanding compilation error with str.contains / String / Pattern](https://users.rust-lang.org/t/solved-need-help-understanding-compilation-error-with-str-contains-string-pattern/22768).

Here's a function that compiles:

```rust
fn compiles() -> bool {
    let word = String::from("hello");
    "hello world".contains(&word)
}
```

But if I omit passing a *reference* to `contains` and pass the actual `String`:

```rust
fn wont_compile() -> bool {
    let word = String::from("hello");
    "hello world".contains(word) // <- missing "&" here before "word"
}
```

, then compilation fails with error

```rust
error[E0277]: expected a `std::ops::FnMut<(char,)>` closure, found `std::string::String`        
  --> src/lib.rs:96:19                                                                          
   |                                                                                            
96 |     "hello world".contains(word)                                                           
   |                   ^^^^^^^^ expected an `FnMut<(char,)>` closure, found `std::string::String`
   |                                                                                            
   = help: the trait `std::ops::FnMut<(char,)>` is not implemented for `std::string::String`    
   = note: required because of the requirements on the impl of `std::str::pattern::Pattern<'_>` for `std::string::String`
```

In this context, as a babbling Rustacean, I had zarroo idea of what rustc was trying to tell me and, assuming the solution (which I found by looking up a `.contains()` example in the Rust book, where a `&` reference was used), my incorrect concluding puzzlement was:

- How a `String` isn't a `std::ops::FnMut<(char,)>`, but  a reference to a `String` is?
- And what is a closure even doing here, where we're all talking `String`s?

As pointed by forum user `krdln`, what's happening is that, as documented by [str.contains](https://doc.rust-lang.org/std/primitive.str.html#method.contains), I should have passed a [Pattern](https://doc.rust-lang.org/std/str/pattern/trait.Pattern.html), whose 6 implementors are:

```rust
impl<'a> Pattern<'a> for char
impl<'a, 'b> Pattern<'a> for &'b str
impl<'a, 'b> Pattern<'a> for &'b String
impl<'a, 'b> Pattern<'a> for &'b [char]
impl<'a, 'b, 'c> Pattern<'a> for &'c &'b str
impl<'a, F> Pattern<'a> for F where F: FnMut(char) -> bool
```

, so rustc's logic behind the error message was:

> *"You’ve passed me something that is not a `char`, `&str`, `&String`, `&[char]` nor `&&str`, so it must be some other type `F`! If you want an `F` to be a pattern, it must implement `FnMut`!"*

## Proposal

When rustc **fails** to compile using the **last** (fallback? generic?) implementor for a trait,

1. Warn that it reached the last implementor.
2. List the missed implementors (always? If short enough? Abbreviated if too long? Only the count if too long? I don't know).

The updated, friendlier error message could look something like that:

```rust
error[E0277]: expected a `std::ops::FnMut<(char,)>` closure, found `std::string::String`        
  --> src/lib.rs:96:19                                                                          
   |                                                                                            
96 |     "hello world".contains(word)                                                           
   |                   ^^^^^^^^ expected an `FnMut<(char,)>` closure, found `std::string::String`
   |                                                                                            
   = help: the trait `std::ops::FnMut<(char,)>` is not implemented for `std::string::String`    
// START ADDITION PROPOSAL =========================================
   = help: reached last generic implementor for `std::str::pattern::Pattern<'_>`, maybe you meant to use one of the 5 other implementors:
           char, &'b str, &'b String, &'b [char], &'c &'b str
// END ADDITION PROPOSAL   =========================================
   = note: required because of the requirements on the impl of `std::str::pattern::Pattern<'_>` for `std::string::String`
```

*<sub>Note: hope the proposal makes sense, or at least you get the intention! I have no idea what I'm doing! This is my first Rust bug! I haven't even finished the Rust book! One last exclamation mark for the road!</sub>*

### PR DESCRIPTION
**Pull request by estebank**
When there are multiple implementors for the same trait that is present
in an unmet binding, modify the E0277 error to refer to the parent
obligation and verify whether borrowing the argument being passed in
would satisfy the unmet bound. If it would, suggest it.

Fix #56368.

### CODE REVIEW

File: src/test/ui/suggestions/issue-62843.stderr [hunk 1]
```diff
@@ -1,11 +1,13 @@
-error[E0277]: expected a `std::ops::FnMut<(char,)>` closure, found `std::string::String`
+error[E0277]: the trait bound `std::string::String: std::str::pattern::Pattern<'_>` is not satisfied
   --> $DIR/issue-62843.rs:4:32
    |
 LL |     println!("{:?}", line.find(pattern));
-   |                                ^^^^^^^ expected an `FnMut<(char,)>` closure, found `std::string::String`
+   |                                ^^^^^^^
+   |                                |
+   |                                expected an `FnMut<(char,)>` closure, found `std::string::String`
+   |                                help: consider borrowing here: `&pattern`
```
Comment (estebank) [inline src/test/ui/suggestions/issue-62843.stderr]: This is the particular common case (`std::str::pattern::Pattern`) that is encountered often by newcomers and this PR is meant to solve.

**Comment by estebank** [comment]
Stable:
Screen Shot 2019-10-15 at 6 45 02 PM

This PR:
Screen Shot 2019-10-15 at 6 44 53 PM

**Comment by estebank** [comment]
It'd be interesting to extend the output with all the implementors of the expected trait, not just the last, and changing the main label to something like `expected something that implements std::str::pattern::Pattern<'_>`.

File: src/test/ui/traits/traits-negative-impls.rs [hunk 2]
```diff
@@ -46,7 +46,7 @@ fn dummy2() {
     impl !Send for TestType {}
 
     is_send(Box::new(TestType));
-    //~^ ERROR `dummy2::TestType` cannot be sent between threads safely
+    //~^ ERROR the trait bound `dummy2::TestType: std::marker::Send` is not satisfied
```
Comment (bjorn3) [inline src/test/ui/traits/traits-negative-impls.rs]: Can this be changed back?

Comment (estebank) [reply src/test/ui/traits/traits-negative-impls.rs]: I can switch the logic to make the `rustc_on_unimplemented` text take precedence again. It seemed safer not to, but let me see what I can do to make it clear.

Comment (estebank) [reply src/test/ui/traits/traits-negative-impls.rs]: done

**Comment by estebank** [comment]
CC @Centril in case you're interested in this.

**Comment by Centril** [comment]
.stderr diffs look good but the code is alien to me :)

File: src/librustc/traits/error_reporting.rs [hunk 3]
```diff
@@ -1234,6 +1249,73 @@ impl<'a, 'tcx> InferCtxt<'a, 'tcx> {
...
+            let found = self_ty.to_string();
+            let new_self_ty = self.tcx.mk_imm_ref(self.tcx.lifetimes.re_static, self_ty);
+            let substs = self.tcx.mk_substs_trait(new_self_ty, &[]);
+            let new_trait_ref = ty::TraitRef::new(obligation.parent_trait_ref.def_id(), substs);
+            let new_obligation = Obligation::new(
+                ObligationCause::dummy(),
+                param_env,
+                new_trait_ref.to_predicate(),
+            );
+            if self.predicate_may_hold(&new_obligation) {
```
[the remark this answers is not in the document]
Comment (estebank) [reply src/librustc/traits/error_reporting.rs]: Done.

**Comment by estebank** [comment]
ping @matthewjasper

**Pull request merged**